apps/aecore/lib/aecore/chain/worker.ex

```diff
@@ -158,25 +179,39 @@ defmodule Aecore.Chain.Worker do
 
   def handle_call({:add_validated_block, %Block{} = new_block, new_chain_state},
                   _from,
-                  %{blocks_map: blocks_map, chain_states: chain_states,
-                    txs_index: txs_index, top_height: top_height} = state) do
+                  %{blocks_map: blocks_map, txs_index: txs_index, 
+                    top_height: top_height} = state) do
     new_block_txs_index = calculate_block_acc_txs_info(new_block)
     new_txs_index = update_txs_index(txs_index, new_block_txs_index)
-    Enum.each(new_block.txs, fn(tx) -> Pool.remove_transaction(tx) end)
     new_block_hash = BlockValidation.block_header_hash(new_block.header)
-
-    updated_blocks_map = Map.put(blocks_map, new_block_hash, new_block)
-    updated_chain_states = Map.put(chain_states, new_block_hash, new_chain_state)
-
+    # refs_list is generated so it contains n-th prev blocks for n-s beeing a power of two. So for chain A<-B<-C<-D<-E<-F<-G<-H. H refs will be [G,F,D,A]. This allows for log n findning of block with given height.
```

Comment:
why is this change needed? the previous code was way simpler

Reply:
I add the `refs` for every block to enable me to traverse chain backwards very fast.

Reply:
Can you explain what's going on here? The usage of arrays looks weird, if working with indexes is really needed, wouldn't an indexed map do the job?